Add support for controlling homekit lights and switches

mjg59 opened the pull request:
## Description:
Add infrastructure for communicating with Homekit IP appliances, along with support for controlling Homekit lightbulbs

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** home-assistant/home-assistant.github.io#4966


## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

mjg59 committed:
Add support for controlling homekit lights and switches

This adds support for controlling lights and switches that expose a HomeKit
control interface, avoiding the requirement to implement protocol-specific
components.

mjg59 commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,173 @@
+            _CONFIGURING[self.hkid] = \
+                    self.configurator.request_config(DOMAIN, self.device_configuration_callback,
+                                            description=description,
+                                            submit_caption="submit",
```
What's the best way to fix this one without ending up with overly long lines?

NovapaX commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,181 @@
+        self._securecon = service['device'].securecon
+        self._aid = service['aid']
+        self._iid = service['service']['iid']
+        self._address = "%s-%s" % (service['device'].hkid, service['aid'])
```
This is used as the unique_id
Is this right? Looking at the Homekit Accessory Protocol I would expect the unique ID to be comprised of the Accessory ID (`aid`?) and the Service or Characteristic ID (`iid`?)
As a single accessory (`aid`) can expose multiple services (light, fan, sensor) at least the `service['iid']` should be included right in the unique_id right?

It appears that `aid` and `iid` can change with a new pairing.  I don't know what `service['device'].hkid` returns. But if that is truly unique and always the same (Mac-address maybe) then it should be enough to just base the unique_id on the `hkid` and the `service['iid']`

Please correct me if I'm wrong :)

mjg59 replied:
Hm, yes - using aid isn't sufficient here, a single accessory can certainly expose multiple lights. Each service should also have a unique serial number, so with luck that'll be stable over pairings.

cdce8p commented:
How does this differ from the already existing HomeKit component and light support?

NovapaX commented:
This will allow to control lights via the Homekit-Accesory-Protocol (HAP). (you could say it will become a 'homekit controller')
Whereas the existing Homekit integration will expose Home-Assistant as a hub (that exposes all kinds of devices) to be used in another 'homekit controller'

arsaboo commented:
Still not sure what is the difference and how will it affect users? Is it replacing the need to have an Apple TV?

NovapaX commented:
No. This is just another way of discovering/controlling devices in Home-Assistant.
So a Light that is Homekit compatible (via WiFi) will be discovered and controllable by Home-Assistant over the Homekit Accessory Protocol.

The problem I see here: What if your light is already configured in Home-Assistant (via another WiFi control protocol, with another component). How is this going to behave?

cdce8p commented:
Please correct me if I'm wrong, but the way I understand it now is that you want to be able to control devices that are only HomeKit compatible with Home Assistant. Whereas the current HomeKit component makes the HA devices accessible for HomeKit.
Basicly something like OpenZWave for HomeKit devices?

mjg59 commented:
Yes, this lets you control any device that's HomeKit compatible without needing device-specific code in Home Assistant.

mjg59 commented:
@NovapaX Yeah that's a good question. The right answer is probably to add a blacklist of devices - using the native support is usually preferable (eg, Tradfri will only expose Ikea bulbs over HomeKit, so the native Tradfri support is necessary if you're using Hue bulbs).

mjg59 commented:
Ouch - the homekit component pulls in a large number library that needs to be built locally, so I guess it should probably go in the commented list for requirements?

mjg59 committed:
Comment out the homekit requirement

This needs to build native code, so leave it commented for now

mjg59 commented:
I've autopep8ed this, but there's still some overly long lines that I can't see a good way to clean up.

balloob commented:
Here is a cool new possibility with this PR: multiple Home Assistant instances working together over the HomeKit protocol

cdce8p commented:
I see your point, but I do not think we should move foreword, at least not that way.
There are a reasons for way I think so, here are a few:
1. Creating jet another HomeKit comonent might lead to additional confusion
2. Introducing another dependency `homekit` in addition to `pyhap` shouldn't be the way to go here, especially since there is an overlap in supported features.
3. From my experience with delveloping the current HomeKit component I know that the protocol used isn't the best nor easiest to implement.
4. All conversion handeling like the one for lights you proposed, should be extracted into a pypi package (except for the parts directly related to Home Assistant).
5. Although we currently use `pyhap`, there isn't any real good package for HomeKit that ideally supports async as well (I know of).
6. As far as I know Apple only provides public information for the accessory side, not the server one. So it might be difficult to reverse engineer the complete server side from it.
7. Due to the limitating nature of HomeKit there are far fewer devices that support only Homekit in contrast to Z-Wave.
8. Lastly, if we say we support lights/devices that support HomeKit, users generally expect that we support all and every features of their devices, which most often simply isn't possible. This is even made harder by the way the HomeKit protocol is designed.

I hope this explains why I think so.

balloob commented:
I think that we should move forward with this. This should be seen as independent of the existing HomeKit integration, because it is a completely different beast. One is for exposing Home Assistant devices, the other one is for integrating devices into Home Assistant. Could the integrations eventually be combined? Maybe, but that's not reason for us to stop this PR today as they are 2 different components.

It's fine that it's not the easiest to implement. We're open source and there are a lot of smart people in this world who like to hack on this. If anything, we should encourage this so that there is an open source example of how to deal with HomeKit.

The fact that some API is not officially documented or that we don't support everything, well even a journey of a thousand miles starts with a single step 😉 This is a great foundation and we can add more features over time. No integration starts with doing everything right at once. We iterate and with each iteration we get more supported, less features etc.

balloob commented:
Also, I disagree that this PR contains too much device specific things.

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,177 @@
+
+    _LOGGER.info("Setting up homekit")
+
+    class hk_device(Entity):
```
Any reason you define this inside the `setup` method and not outside of it?

mjg59 replied:
Fixed

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,177 @@
+            self.config = config
+            self.configurator = hass.components.configurator
+
+            data_dir = os.path.join(get_default_config_dir(), homekit_dir)
```
Use `hass.config.path(homekit_dir)` to use the current config dir that is in use

mjg59 replied:
Fixed

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,173 @@
+
+        def get_serial(self, accessory):
+            for service in accessory['services']:
+                if homekit.ServicesTypes.get_short(service['type']) != 'accessory-information':
```
```python
if (homekit.ServicesTypes.get_short(service['type']) !=
        'accessory-information'):
    continue
```

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,173 @@
+                    continue
+                aid = accessory['aid']
+                for service in accessory['services']:
+                    s = {'type': homekit.ServicesTypes.get_short(service['type']),
```
```python
s = {
    'type': …
}
```

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,177 @@
+            description = """Please enter the Homekit code for your {}
+                          """.format(self.model)
+            _CONFIGURING[self.hkid] = \
+                self.configurator.request_config(DOMAIN, self.device_configuration_callback,
```
```python
_CONFIGURING[self.hkid] = self.configurator.request_config(
    DOMAIN, self.device_configuration_callback,
    description=description,
)
```

cdce8p commented:
@balloob It might have been my previous experience with the HomeKit coponent and it's issues, from which I wrote that comment. If it works in the end, it will be worth it. My concern was getting there 😉

Unfortunately I can't really add much to it, since I don't have access to any real HomeKit device.
Just one thing to keep in mind: The HomeKit component exposes a bridge accessory that shouldn't then be discovered by Home Assistant. That could get wired.

maddox commented:
This is huge. This means any devices that are HomeKit only and don't have any other way to integrate with Home Assistant, can now be integrated....and with LOCAL CONTROL AND PUSH UPDATES.

This could also potentially be a big work around to platforms that are only clouded. For example, Ecobee. Right now it's a polling platform that isn't very reliable in terms of latency. Also, it has to hit their cloud to do any control.

But in HomeKit, the Ecobee is essentially a local device. You could ditch your Home Assistant Ecobee config, and just use the sensors and tstat that HomeKit exposes and have local control and push updates for these components. 

Glorious.

mjg59 commented:
@maddox Note that (right now) the implementation doesn't handle events, but I'll look into that

maddox commented:
🆒 

It doesn’t seem mergable unless it does though. Having state is really important. Omnidirectional control with 1st class components inside the Home Assistant (which they would be after bridging over from HomeKit) domain would not fulfill expectations. 

Like, if a light shows up in Home Assistant from HomeKit, users will expect its state to be trustable and not just that they can control it. 

Also it would be half as useful 😜

Really excited about the opportunities that this spike has for the platform! Thanks!

marthoc commented:
@mjg59 Awesome work on this, this is very exciting!

mjg59 commented:
@houndci-bot you are my worst friend

mjg59 commented:
As an example of why this is useful - the version I just pushed makes the Vocolink RGB bulbs and switches that are relatively cheap on Amazon right now work out of the box without any custom protocol code. As vendors continue adding support for HomeKit in order to work with Siri this is going to be the easiest way to get new devices working without additional code. There's still a fair amount to do here to satisfy review comments, but I hope it seems like it's useful :)

balloob commented:
I'm very stoked about this PR. It's cool that Apple has managed to create a local protocol that manufacturers are willing to adopt.

I'll do a first pass review just to help get this PR moving. Could you please not add any more types of supported devices in this PR? Let's get this PR merged first. It's already at 400+ lines.

balloob commented on .coveragerc:
```diff
@@ -404,6 +404,7 @@ omit =
     homeassistant/components/feedreader.py
     homeassistant/components/foursquare.py
     homeassistant/components/goalfeed.py
+    homeassistant/components/homekit_controller.py
```
Since we have light and switch platforms now, let's move it to the top of the `coveragerc` file and make the platform matcher use a wildcard (`*`)

mjg59 replied:
Done

balloob commented on homeassistant/components/discovery.py:
```diff
@@ -55,6 +56,7 @@
     SERVICE_DECONZ: ('deconz', None),
     SERVICE_DAIKIN: ('daikin', None),
     SERVICE_SAMSUNG_PRINTER: ('sensor', 'syncthru'),
+    SERVICE_HOMEKIT: ('homekit_controller', None),
```
So I want to treat this one with caution. If the next release of Home Assistant all of a sudden starts discovering devices that are already added to Home Assistant via their own API, people will be pissed. Especially because right now there is no way to disable it without disabling all of a specific discovery type.

This will be less of an issue if homekit_controller would use the new config entries support instead of the configurator, as it won't trigger new states in the UI. (video of it in action [here](https://www.youtube.com/watch?v=_qpKe-FW9Yk), more info on config entries [here](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/config_entries.py#L1), example config flow [here](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/hue/config_flow.py#L43)). Config entries are currently behind a feature flag (add `config_entries: 1` under `config:`) but will be available for all in the next release.

However, to not blow this PR up any further, instead of converting from using configurator to config entries, you can just change the discovery component to make `SERVICE_HOMEKIT` an opt-in discovery. That way, a future PR can convert homekit controller to config entries.

mjg59 replied:
Not quite clear on what you mean by opt-in discovery here - the only configuration that discovery seems to have is to ignore things. Should I be adding additional config support for explicitly enabling devices?

balloob replied:
I think that would be a good idea. I wouldn't mind to tie it directly to homekit.

```yaml
discovery:
  discover_homekit: true  # default is False
```

(I expect this feature to be removed when someone migrates homekit_controller to config entries)

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,178 @@
+homekit_dir = 'homekit'
+DOMAIN = 'homekit_controller'
+
+_CONFIGURING = {}
```
Please use `hass.data` to store global data. It's a dictionary in which you can store anything you want.

mjg59 replied:
Done

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,178 @@
+    'outlet': 'switch',
+}
+
+KNOWN_ACCESSORIES = []
```
These should also be moved to `hass.data`. Use any key you want in the dict but since it is shared among all components, prefix your key with your domain.

mjg59 replied:
Done

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,178 @@
+
+REQUIREMENTS = ['homekit==0.5']
+
+homekit_dir = 'homekit'
```
Since we don't want users to edit files in this dir, let's prefix the name with a `.`

mjg59 replied:
Done

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+            if ctype == "on":
+                self._chars['on'] = characteristic['iid']
+                self._on = characteristic['value']
+            if ctype == 'brightness':
```
`elif`?

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+                                                   self._saturation / 100,
+                                                   self._brightness / 100)
+            return (red * 255, green * 255, blue * 255)
+        return (255, 255, 255)
```
We should return `None` if no color.

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+        """Turn the specified light on."""
+        self._state = True
+
+        rgb = kwargs.get(ATTR_RGB_COLOR)
```
While this PR was in progress, we've switched the internal color representation to HS, so you should just be able to get `ATTR_HS_COLOR`

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+                            'value': False}]
+        body = json.dumps({'characteristics': characteristics})
+        self._securecon.put('/characteristics', body)
+        self.update_state()
```
You have should_poll set to True. That means that Home Assistant will call `update()` after it has called a service. So you can remove this statement if you rename the `update_state` to `update`

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+    def update_state(self):
+        response = self._securecon.get('/accessories')
+        data = json.loads(response.read().decode())
+        for accessory in data['accessories']:
```
This seems very inefficient, no? Each entity will fetch the state for all accessories?

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+                if service['iid'] != self._iid:
+                    continue
+                self.update_characteristics(service['characteristics'])
+                self.schedule_update_ha_state()
```
How often can this be called? Can we`break` here? Else we should move this call down and make sure we only call it once.

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+        """Return the brightness of this light between 0..255."""
+        if self._features & SUPPORT_BRIGHTNESS:
+            return (self._brightness * 255 / 100)
+        return self._on * 255
```
return `None` if we don't support brightness.

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+        """Return the color temperature."""
+        if self._features & SUPPORT_COLOR_TEMP:
+            return self._color_temperature
+        return 3000
```
Return `None`

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+    @property
+    def assumed_state(self):
+        """Return the assumed state."""
+        return True
```
The state doesn't seems like it is assumed but is read from Homekit?

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+
+    def turn_on(self, **kwargs):
+        """Turn the specified light on."""
+        self._state = True
```
This is not necessary if we update after sending this command?

balloob commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, service):
```
Let's keep the method signature the same as other platforms:
 - rename `service` to `discovery_info`
 - default `discovery_info` to `None`

You will also want to return from the method if discovery info is None, in case someone puts the platform config in their config file.

mjg59 replied:
Done

balloob commented on homeassistant/components/switch/homekit_controller.py:
```diff
@@ -0,0 +1,98 @@
+_LOGGER = logging.getLogger(__name__)
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, service):
```
Same comment as for light.

mjg59 replied:
Done

balloob commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,178 @@
+                    _LOGGER.info("Found %s", s['type'])
+                    component = HOMEKIT_ACCESSORY_DISPATCH.get(s['type'], None)
+                    if component is not None:
+                        discovery.load_platform(hass, component, DOMAIN, s,
```
Discovery info is only allowed to contain JSON serializable values as it leverages the event bus. So instead of sending all info over, send over the aid/service combo and fetch it from `hass.data` inside the platform setup.

mjg59 replied:
Done

balloob commented on homeassistant/components/switch/homekit_controller.py:
```diff
@@ -0,0 +1,98 @@
+        self._chars = {}
+        self.update_characteristics(service['service']['characteristics'])
+
+    def update_state(self):
```
Same comments as for light:
 - rename to `update`
 - Can we cache the call to `/accessories` between all devices?
 - Call `schedule_update_ha_state` once

mjg59 replied:
Largely done

balloob commented on homeassistant/components/switch/homekit_controller.py:
```diff
@@ -0,0 +1,98 @@
+                            'value': True}]
+        body = json.dumps({'characteristics': characteristics})
+        self._securecon.put('/characteristics', body)
+        self.update_state()
```
Can be removed if renamed to `update`

mjg59 replied:
Done

MartinHjelmare reviewed (changes requested):
Cool!

mjg59 commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,178 @@
+    self.send(msg)
+
+
+# pylint: disable=unused-argument, too-many-function-args
```
Fixed

mjg59 commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,178 @@
+    """Set up for Homekit devices."""
+    import homekit
+
+    _LOGGER.info("Setting up homekit")
```
Done

mjg59 commented on homeassistant/components/homekit_controller.py:
```diff
@@ -0,0 +1,178 @@
+            data_dir = os.path.join(get_default_config_dir(), homekit_dir)
+            if not os.path.isdir(data_dir):
+                os.mkdir(data_dir)
+            self.pairing_file = os.path.join(data_dir, "hk-" + hkid)
```
Done

mjg59 commented on homeassistant/components/switch/homekit_controller.py:
```diff
@@ -0,0 +1,98 @@
+        return self._on
+
+    @property
+    def should_poll(self):
```
Done

mjg59 commented on homeassistant/components/switch/homekit_controller.py:
```diff
@@ -0,0 +1,98 @@
+
+    def turn_on(self, **kwargs):
+        """Turn the specified switch on."""
+        self._on = True
```
Done

mjg59 commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+    ATTR_BRIGHTNESS, ATTR_RGB_COLOR, ATTR_COLOR_TEMP, SUPPORT_BRIGHTNESS,
+    SUPPORT_RGB_COLOR, SUPPORT_COLOR_TEMP, Light)
+
+REQUIREMENTS = ['homekit==0.5']
```
Done

mjg59 commented on homeassistant/components/switch/homekit_controller.py:
```diff
@@ -0,0 +1,98 @@
+
+from homeassistant.components.switch import SwitchDevice
+
+REQUIREMENTS = ['homekit==0.5']
```
Done

mjg59 commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,182 @@
+class HomekitLight(Light):
+    """Representation of a Homekit light."""
+
+    def __init__(self, service):
```
Done

MartinHjelmare commented:
Please don't squash after code review has started. Squashing makes it a lot harder to follow the changes.

MartinHjelmare commented on homeassistant/components/homekit_controller/__init__.py:
```diff
@@ -0,0 +1,215 @@
+        self._address = "homekit-{}-{}".format(devinfo['serial'], self._iid)
+        self._features = 0
+        self._chars = {}
+        self.update()
```
Remove this. Instead call `add_devices` with a second aro`True`, in the platforms:
```py
add_devices(devices, True)
```

MartinHjelmare commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,136 @@
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up Homekit lighting."""
+    if discovery_info is not None:
+        add_devices([HomeKitLight(hass, discovery_info)])
```
Add an argument `True` here.

MartinHjelmare commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,136 @@
+                self._saturation = characteristic['value']
+
+    @property
+    def unique_id(self):
```
You can remove this property since it's already defined in the parent.

MartinHjelmare commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,136 @@
+        return self._address
+
+    @property
+    def name(self):
```
Same as above.

MartinHjelmare commented on homeassistant/components/light/homekit_controller.py:
```diff
@@ -0,0 +1,136 @@
+    @property
+    def assumed_state(self):
+        """We return the actual state."""
+        return False
```
This is the default so you can remove this property.

MartinHjelmare commented on homeassistant/components/switch/homekit_controller.py:
```diff
@@ -0,0 +1,58 @@
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up Homekit switch support."""
+    if discovery_info is not None:
+        add_devices([HomeKitSwitch(hass, discovery_info)])
```
Add `True`.

MartinHjelmare commented on homeassistant/components/homekit_controller/__init__.py:
```diff
@@ -0,0 +1,215 @@
+class HomeKitEntity(Entity):
+    """Representation of a Home Assistant HomeKit device."""
+
+    def __init__(self, hass, devinfo):
```
I would prefer if you pass in only the `accessory`. That makes it more clear what is needed.

MartinHjelmare reviewed (commented):
Very nice!

MartinHjelmare reviewed (commented):
Add a stub method in `HomeKitEntity` for `update_characteristics`:
```py
def update_characteristics(self):
    raise NotImplementedError
```

Add init methods in the child platform classes that only inits the child specific attributes:
```py
def __init__(self, *args):
    super().__init__(*args)
    self._on = None
    self._brightness = None
    ...
```

mjg59 committed:
Review updates

mjg59 committed:
Make HomeKit auto-discovery optional

Add an "enable" argument to the discovery component and add a list of
optional devices types (currently just HomeKit) to discover

mjg59 committed:
Further review comments

mjg59 committed:
Update requirements_all.txt

mjg59 committed:
Fix houndci complaints

mjg59 commented:
Ok, I think the current set of review feedback is satisfied - anything else needed?

balloob commented:
Deleted a bunch of Hound comments. The PR kept timing out on GitHub when trying to open it.

balloob commented on homeassistant/components/homekit_controller/__init__.py:
```diff
@@ -0,0 +1,218 @@
+
+KNOWN_ACCESSORIES = "{}-accessories".format(DOMAIN)
+
+KNOWN_DEVICES = {}
```
Please use `hass.data` to store "global" data. We should not mutate constants.

balloob commented on homeassistant/components/discovery.py:
```diff
@@ -103,6 +111,12 @@
+    enabled_platforms = config[DOMAIN][CONF_ENABLE]
+
+    for service in enabled_platforms:
+        SERVICE_HANDLERS[service] = OPTIONAL_SERVICE_HANDLERS[service]
```
We should not mutate constants. Better to check inside `new_service_found`:

```python
if (not comp_plat and service in OPTIONAL_SERVICE_HANDLERS and
        service in enabled_platforms):
    comp_plat = OPTIONAL_SERVICE_HANDLERS[service]
```

balloob commented on .coveragerc:
```diff
@@ -415,6 +418,7 @@ omit =
     homeassistant/components/light/decora.py
     homeassistant/components/light/flux_led.py
     homeassistant/components/light/greenwave.py
+    homeassistant/components/light/homekit_controller.py
```
This line is already covered by the above wildcard line.

balloob commented on homeassistant/components/homekit_controller/__init__.py:
```diff
@@ -0,0 +1,218 @@
+KNOWN_ACCESSORIES = "{}-accessories".format(DOMAIN)
+
+KNOWN_DEVICES = {}
+DEVICES = []
```
Same

balloob commented:
Only a few very minor comments left. Also Travis did not run, not sure what is up there. Maybe because of the merge conflict.

MartinHjelmare commented on homeassistant/components/homekit_controller/__init__.py:
```diff
@@ -0,0 +1,218 @@
+        """Return the name of the device if any."""
+        return self._name
+
+    def update_characteristics(self):
```
Make sure this has the same parameters as the methods that you need to implement in the child classes, ie add `characteristics`.

MartinHjelmare commented on homeassistant/components/switch/__init__.py:
```diff
@@ -134,7 +134,6 @@ def toggle(hass, entity_id=None):
 
 class SwitchDevice(ToggleEntity):
     """Representation of a switch."""
-
```
Please revert this change.

mjg59 replied:
Done, not sure why it's not flagging this.

MartinHjelmare replied:
Doesn't seem to be done. 😉

MartinHjelmare commented on .coveragerc:
```diff
@@ -106,6 +106,9 @@ omit =
     homeassistant/components/hive.py
     homeassistant/components/*/hive.py
 
+    homeassistant/components/homekit_controller.py
```
Please update the path to point at the `__init__.py` file under `homekit_controller` component directory.

mjg59 committed:
Further review updates

mjg59 commented:
Should I rebase this, or leave it up to you?

MartinHjelmare commented:
Please solve the merge conflicts.

mjg59 commented:
Rebased, let me know!

mjg59 committed:
Final review fixup

mjg59 committed:
Lint fixups

mjg59 committed:
Fix discovery tests

mjg59 committed:
Further review updates

MartinHjelmare reviewed (approved):
Awesome!

maddox commented:
danggg. Nice work on this everyone!

elbogi commented:
There's an issue posted (#14062) regarding this new component, is this something that's addressed in the latest RC?
